openedx/core/djangoapps/content/learning_sequences/views.py

```diff
@@ -163,35 +165,60 @@ def get(self, request, course_key_str, format=None):  # lint-amnesty, pylint: di
             raise NotFound() from does_not_exist_err
 
         serializer = self.UserCourseOutlineDataSerializer(user_course_outline_details)
         return Response(serializer.data)
 
-    def _determine_user(self, request):
+    def _determine_user(self, request, course_key: CourseKey) -> User:
         """
-        Requesting for a different user (easiest way to test for students)
-        while restricting access to only global staff. This is a placeholder
-        until we have more full fledged permissions/masquerading.
+        For which user should we get an outline?
+
+        Uses a combination of the user on the request object and a manually
+        passed in "user" parameter. Ensures that the requesting user has
+        permission to view course outline of target user. Raise request-level
+        exceptions otherwise.
+
+        The "user" querystring param is expected to be a username, with a blank
+        value being interpreted as the anonymous user.
         """
-        requested_username = request.GET.get("user")
+        target_username = request.GET.get("user")
 
-        # Simple case: No username passed in, so it's just the request.user
-        if not requested_username:
+        # Sending no "user" querystring param at all defaults us to the user who
+        # is making the request.
+        if target_username is None:
             return request.user
 
-        # If we're here, then the requesting user is asking for someone else's
-        # course outline. Right now, only global staff is allowed to do that.
-        if request.user.is_staff:
-            try:
-                user = User.objects.get(username=requested_username)
-                return user
-            except User.DoesNotExist as err:
-                raise NotFound("User {requested_username} does not exist.") from err
-
-        raise PermissionDenied(
-            "User {request.user.username} does not have permission to view course outline as {requested_username}"
-        )
+        # Users can always see the outline as themselves.
+        if target_username == request.user.username:
+            return request.user
+
+        # Otherwise, do a permission check to see if they're allowed to view as
+        # other users.
+        if not can_see_content_as_other_users(request.user, course_key):
+            display_username = "the anonymous user" if target_username == "" else target_username
+            raise PermissionDenied(
+                f"User {request.user.username} does not have permission to "
+                f"view course outline for {course_key} as {display_username}."
+            )
+
+        # If we've gotten this far, their permissions are fine. Now we handle
+        # the masquerade use case...
+
+        # Having a "user" querystring that is a blank string is interpreted as
+        # "show me this outline as an anonymous user".
+        if target_username == "":
+            return AnonymousUser()
```

Comment:
Question: Is this logic piece also used elsewhere for masquerading? It'd be nice to encapsulate all this logic in a custom Django user - or just a RequestingUser object constructor used universally.

Reply:
Not that I'm aware of. It's a new convention I'm trying out. I agree that we need more uniform conventions (probably an OEP), but I don't have the time and energy to push that forward right now.